### PR TITLE
GH-14: Make digest (and other key options) configurable.

### LINKED ISSUE #14: Make password encoding configurable.
Currently we use `sha1` encoding, which is not optimal.  We also don't provide any means to configure this.  This issue will cover:

1. Updating to use a new default encoding.
2. Making it possible and testing configuring this in real time.
3. Testing working with records that use the legacy `sha1` encoding when that is no longer the default.

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
Fixes #14.

### CODE REVIEW

**Comment by the-t-in-rtf** [comment]
Hi, All.  @BlueSlug and @jobara have been discussing improving the range of digests available for password encryption, and I put together a pull to accomplish this.  I am copying in @amb26 and hoping he has a small amount of bandwidth, as this is a major release level change which, in addition to making it possible to configure the digest, removes the requirement to interoperate with CouchDB records.  There is a migration guide in the updated README.

I would propose that @amb26 review the code.  Although @BlueSlug and @jobara are also welcome to comment on the code changes, I was hoping they could comment on the "fitness for purpose" by trying it in the downstream project where they hoped to used the feature, and then commenting here.  With this in mind, I have cut a dev release (`2.0.0-dev.20201127T130239Z.e5fe1aa.GH-14`) for them to use.

I am also copying @sgithens in case there is any remaining work he's done that would be affected.

**Comment by the-t-in-rtf** [comment]
@BlueSlug, I left the default digest encoding set to `sha256`, I welcome comments on whether that's a good choice.

File: src/js/server/reset.js [hunk 1]
```diff
@@ -78,6 +78,18 @@ fluid.express.user.reset.handler.checkResetCode = function (that, dataSourceResp
 
 fluid.defaults("fluid.express.user.reset.handler", {
     gradeNames:  ["fluid.express.handler"],
+    digest: "{fluid.express.user.reset.post}.options.digest",
```
Comment (amb26) [inline src/js/server/reset.js:81]: Rather than repeating this material in each handler and also subcomponent definition, why not factor it, and also bindings to it, into a fluid.express.user.passwordEncryptOptions grade?

Comment (the-t-in-rtf) [reply src/js/server/reset.js]: Good idea.  I made a pair of grades, one that "holds" the options, another that "consumes" them.

File: src/js/server/api.js [hunk 2]
```diff
@@ -104,7 +106,7 @@ fluid.defaults("fluid.express.user.api", {
         reset: {
             type: "fluid.express.user.reset",
             options: {
-                priority: "after:session"
+                priority:   "after:session"
```
Comment (jobara) [inline src/js/server/api.js:109]: Is this whitespace change intentional?

**Comment by jobara** [comment]
@the-t-in-rtf it appears you missed updating the line that sets the document `_id` with the couch db prefix. https://github.com/the-t-in-rtf/fluid-express-user/blob/d6f937e8d8be1197e6b696f44825be06cad85b7e/src/js/server/utils.js#L109 
Ideally the id would be something configurable so that if we want to set it as something other than the username we could. For example we could use some other unique id that way the user would have the ability to change their username, e-mail, and etc after account creation.

File: src/js/server/api.js (hunk 2 above)
Comment (the-t-in-rtf) [reply src/js/server/api.js:109]: it was indented with its previous neighbours that I removed.  I'll update it.

**Comment by the-t-in-rtf** [comment]
> Ideally the id would be something configurable

@jobara, I updated to make nearly every bit of a new user record configurable in the same transform.  There is one field I had difficulty updating in a single transform, other than that everything that can be configured is configurable via options.

I cut a new dev release with the updated, work, it's `2.0.0-dev.20201130T162948Z.dd089e1.GH-14`.

**Comment by the-t-in-rtf** [comment]
I also fixed the goof with the leftover prefix, which was also in the tests, otherwise we'd have heard about it before.

**Comment by BlueSlug** [comment]
> @BlueSlug, I left the default digest encoding set to `sha256`, I welcome comments on whether that's a good choice.

@the-t-in-rtf: after a bit of reading, `sha256` is certainly sufficient for the time being. That said, there are concerns about how, given the digest of a term `x`, one could discover the digests of terms that begin with `x`. `sha3-256` would be more secure but likely slower, though the things I have read about the `blake` family of algorithms are promising. Here's a bit more info on Blake: https://www.blake2.net/

TL;DR: My recommendation for the default is `blake2b512` or `blake2s256`, given the list from `(require ("crypto")).getHashes()`

**Comment by the-t-in-rtf** [comment]
Thanks @BlueSlug.  Hopefully they'll have [blake3](https://github.com/BLAKE3-team/BLAKE3) support in node@16. For now, I've set the default to `blake2b512`.  I cut a new dev release with the updated defaults, it's `2.0.0-dev.20201201T093241Z.c2278e5.GH-14`.

File: src/js/server/lib/password.js [hunk 3]
```diff
@@ -17,9 +14,9 @@ var crypto = require("crypto");
 fluid.registerNamespace("fluid.express.user.password");
 fluid.express.user.password.encode = function (password, salt, iterations, keyLength, digest) {
     // Set defaults that are useful in dealing with CouchDB and express-couchUser data.
-    iterations = iterations || 10;
-    keyLength  = keyLength  || 20;
-    digest     = digest     || "sha1"; // Already used when the value is omitted, but specified for future-proofing.
+    iterations = iterations !== undefined ? iterations : 10;
+    keyLength  = keyLength !== undefined ? keyLength : 20;
+    digest     = digest !== undefined ? digest : "sha256";
```
Comment (jobara) [inline src/js/server/lib/password.js:19]: keyLength doesn't show up in the user record, is that intentional or does it need to be there as well?

Comment (the-t-in-rtf) [reply src/js/server/lib/password.js:19]: It doesn't.  You can tell the length from the key itself.  Take a look at all of the test records, which don't have it, but can still be used to log in.

Comment (the-t-in-rtf) [reply src/js/server/lib/password.js:19]: In other words, you need to know how long to make the key before you make it, once you make it, you can just see how long it is.

Comment (jobara) [reply src/js/server/lib/password.js:19]: @the-t-in-rtf thanks for the clarification, that makes sense.

**Comment by jobara** [comment]
@the-t-in-rtf I forgot to mention that I'm using the dev release in my [SJRK-405 branch](https://github.com/jobara/sjrk-story-telling/tree/SJRK-405). Things seem to be working. Thanks.

**Comment by the-t-in-rtf** [comment]
@amb26, if you're happy with this, I'd like to merge and cut a full release to support the work on the SJRK.

**Comment by jobara** [comment]
@the-t-in-rtf I'm not sure if this fits with this PR, but one other issue I ran into was related to [`fluid.express.user.utils.unlockUser`](https://github.com/the-t-in-rtf/fluid-express-user/blob/c2278e52586b199dcaca2a3485bb4fd1a77bacfe/src/js/server/utils.js#L181-L204). In the storytelling tool we are using the e-mail address for the username, so we don't actually need the username property at all. However, this function assumes the username based on how it [parses](https://github.com/the-t-in-rtf/fluid-express-user/blob/c2278e52586b199dcaca2a3485bb4fd1a77bacfe/src/js/server/utils.js#L185) the response from the call to `that.byUsernameOrEmailReader.get`. To work around that I just mapped the email onto the `username` as well in the rules configuration. However, it would be nice to not have to duplicate the entry and be able to use `fluid.express.user.utils.unlockUser` with either a username or email.

**Comment by the-t-in-rtf** [comment]
I saw those chats, but I really think we should handle it in a separate ticket, as we need to make it properly configurable.

File: src/js/server/utils.js [hunk 4]
```diff
@@ -49,14 +49,30 @@ fluid.defaults("fluid.express.user.utils", {
...
         createUserWrite: {
-            "name":          "username", // Default rules are designed to cater to CouchDB  and express-couchUser conventions, but can be overriden.
-            "username":      "username",
-            "email":         "email",
-            roles:           { literalValue: []},
-            type:            { literalValue: "user"},
-            password_scheme: { literalValue: "pbkdf2"},
-            iterations:      { literalValue: 10},
-            verified:        { literalValue: false}
+            // TODO: Discuss w/ Antranig why this isn't equivalent to:
```
Comment (amb26) [inline src/js/server/utils.js:52]: Transforms don't support expansion in the members "inputPath" and "outputPath" - these are always interpreted as literal paths or path segments

Comment (the-t-in-rtf) [reply src/js/server/utils.js:52]: Thanks.
